ZenPacks/zenoss/OpenStackInfrastructure/zenopenstack.py

```diff
@@ -56,6 +56,76 @@
 pb.setUnjellyableForClass(OpenStackDataSourceConfig, OpenStackDataSourceConfig)
 
 
+class CeilometerMessageCache(object):
+    """
+    Process payload data that looks like following:
+
+        [{u'event_type': u'volume.update.end',
+        u'generated': u'2019-01-03T14:47:05.300175',
+        u'message_id': u'239a077f-910b-4131-926d-460c4d81da58',
+        u'raw': {},
+        u'traits': [[u'status', 1, u'available'],
+                    [u'user_id', 1, u'4b3448a3057c411d827de9bfb77a5599'],
+                    [u'service', 1, u'volume.osi-p'],
+                    [u'availability_zone', 1, u'nova'],
+                    [u'tenant_id', 1, u'313da8e3ab19478e82be9c50e6b1a04b'],
+                    [u'created_at', 1, u'2018-12-13T17:02:14+00:00'],
+                    [u'resource_id', 1, u'375666bf-833a-4102-972e-402031f47f41'],
+                    [u'host', 1, u'osi-p@lvm#lvm'],
+                    [u'request_id', 1, u'req-b34e6276-5fae-4914-9736-60fe259c24db'],
+                    [u'display_name', 1, u'junkVolume2'],
+                    [u'project_id', 1, u'313da8e3ab19478e82be9c50e6b1a04b'],
+                    [u'type', 1, u'45bbf67b-c222-4bc6-a3a1-97804d9b1d53'],
+                    [u'size', 1, u'1']]}]
+    """
+
+    def __init__(self):
+        # Create a dict cache that consists of deques max size 100:
+        self.cache = defaultdict(partial(deque, maxlen=100))
+
+    def store_message(self, request):
+        """Store events in the cache, update statistics.
+
+        Hostname formatted as <host>:<port>/<uri> . For example:
+        '10.1.1.12:8242/ceilometer/v1/samples/osi_p'
+
+        This allows for separate samples and events cache data.
+        """
+
+        # Create a unique host_url that has: host:port/uri
+        host_port = request.getAllHeaders().get('host')
+        host_url = host_port + request.uri
+        message = request.content.getvalue()
+
+        # Add the host's message to the cache
+        if host_url:
+            self.cache[host_url].append(message)
+
+        # Now use Metrology to record the event
+        Metrology.meter(host_url).mark()
```

Comment:
I would suggest building a bit more information into the instrument names to make them somewhat more self-describing.   (Just in case we decide to send them somewhere else in the future or something)

For example, 
http.post.ceilometer.v1.samples.[somedeviceid].requests    

Try to have a convention that facilitates other instrument names we might want to capture, like
http.requests (total requests)
http.post.requests (total posts)
http.post.ceilometer.v1.samples.requests   (samples from all devices)

Something like that?